### ISSUE TITLE
TRN-1 — Staged training pipeline with typed stage contracts and an explicit training loop

### ISSUE THREAD

**Issue by aacostadiaz** [ACEsuit/mace#1575]
**Depends on:** CFG-1 (#1574), DATA-3 (#1571), ARCH-3 (#1562) · **Blocks:** TRN-2 (#1576), TRN-3 (#1577), TRN-4 (#1578), TRN-5 (#1639), CLI-1 (#1579), FT-1 (#1580), ELEC-1 (#1591), GATE-3 (#1585)

**Context:** Today's training driver is `mace/cli/run_train.py` — 1,249 lines, with `run(args)` a single function spanning lines 89→1245 (~1,157 lines). The argparse `Namespace` is the data model: `run()` reads **74 distinct `args.*` attributes** and **assigns to `args` in place at 43 sites** (`args.heads = prepare_default_head(args)` at `:276`, `args.r_max = model_foundation.r_max.item()` at `:212`, `args.loss = "universal"` at `:435`, `args.avg_num_neighbors = get_avg_num_neighbors(...)` at `:835`; foundation presets overwrite user inputs at `:247-249` unless `--force_mh_ft_lr`). Model identity is a string that rewrites the observable set: `:588-625` branches on the model-class name and hand-sets six `args.compute_*` booleans. Everything — foundation resolution, head prep, data loading, E0s, dataloaders, `configure_model`, in-place cueq/oeq conversion, optimizer/scheduler, SWA, EMA, resume, the LBFGS optimizer swap, `tools.train(...)` — is interleaved in one scope with no stage boundary; the matching parser (`mace/tools/arg_parser.py`) is 1,417 lines of untyped flags. This ticket replaces `run()` with an **orchestrator of ~40 lines** wiring three typed stages, plus an **explicit train/validate/checkpoint/EMA/stage-two loop** (a few hundred readable lines, reproducible by a power user as a script) — heavy training abstractions (PyTorch Lightning, Accelerate) are explicitly out, and so are callback/hook frameworks; loops stay explicit and simple. Legacy (`mace/tools/train.py`, 815 lines: `train()` `:156`, `train_one_epoch` `:369`, `take_step` `:421`, `take_step_lbfgs` `:467`, `evaluate()` `:581`, `MACELoss` `:614`, SWA container `:42`, the stage-two swap at `:229-243`) is read in place as the behavioural oracle; reproduce its numbers, never its structure. TRN-1 (#1575) is on the critical path: every remaining Phase 3 ticket builds on its stage seams.

**Interface & constraints:**

1. **The pipeline** — each stage takes and returns a typed object (Pydantic / dataclass, boundary types defined in `mace_core`); the argparse namespace never crosses a stage boundary:

```
ResolvedConfig ──▶ DataStage ──▶ DataBundle ──▶ ModelStage ──▶ BuiltModel ──▶ TrainStage ──▶ TrainedModel
   (CFG-1 (#1574))          (TRN-1 (#1575))      (typed)        (TRN-1 (#1575))        (typed)         (TRN-1 (#1575))       (+ metadata)
```

2. **Responsibility map** (legacy `run()` block → its stage):

| Legacy responsibility (frozen `mace/cli/run_train.py`) | v1 home |
|---|---|
| Data loading, xyz/h5/aselmdb branching + ASE-file splitting (`:323-466`, dataset assembly `:637-831`) | `DataStage` — dataset backends yield `Configuration`s (DATA-1 (#1569)/2), graph building/collation once above the backend (DATA-3 (#1571)) |
| z-table (`:468-498`) + atomic energies + E0s + foundation padding (`:500-635`) | `DataStage` — typed `E0Spec` resolved via `resolve_e0s` (see 3); every keep-going fallback becomes a hard error |
| Pseudolabels (`:645-659`) | a **callable stage seam** in `DataStage` (implementation lands in FT-2 (#1581)), not `requires_grad` juggling welded into training |
| `args.compute_*` set by model-class string (`:588-625`) | gone — observables are **declared in config**; declaring `energy, forces, dipole` creates the heads (ARCH-3 (#1562), CORE-1 (#1555)) |
| `configure_model` (`:838`) | `ModelStage`: `ResolvedConfig` + statistics → two-layer `BaseMACE` + `MACEOutputs` |
| In-place cueq/oeq conversion (`:875-906`, incl. the mutual-exclusion fixup `args.enable_oeq = False` at `:880`) | gone — backend dispatch resolves once at model build time (BKD-1 (#1560)) |
| Optimizer / LRScheduler / SWA / EMA / resume (`:903-943`) | `TrainStage` setup; per-stage schedules (schedule algebra itself is TRN-2 (#1576), the LBFGS stage kind is TRN-5 (#1639)) |
| `tools.train(...)` + checkpoint + save (`:1002…`, `torch.save(model_to_save, model_path)` at `:1183`) | the explicit loop; checkpoints carry the resolved config as metadata (safetensors format + DDP port are TRN-4 (#1578)) |
| Foundation-model resolution + preset overrides (`:207-274`), head prep / `pt_head` (`:275-322`) | `finetune` config section + typed head configs (CFG-1 (#1574), FT-1 (#1580)) — presets are declared defaults, never post-parse mutations |

3. **E0 resolution is called exactly once, before model build**, via the pure `mace_core` function (torch-free; `"estimated"` is the only variant needing an injected `EnergyPredictor` from `mace_torch`):

```python
def resolve_e0s(
    spec: E0Spec, z_table: AtomicNumberTable,
    train_configs: Sequence[Configuration] | None, *,
    foundation_e0s: Mapping[int, float] | None = None,
    predict_energy: EnergyPredictor | None = None,     # required iff kind == "estimated"
) -> tuple[dict[int, float], E0sProvenance]: ...
```

Every element of the z-table must be resolved or the function raises — replacing today's spread of `ValueError`, the warn-and-use-`0.0` path for an unlabelled `IsolatedAtom` (`mace/data/utils.py:333-337`), the `LinAlgError → all zeros` path (`:381-387`), and the `head_energies.get(z, 0.0)` zero-padding under `--foundation_model_elements` (`mace/cli/run_train.py:578-586`). Provenance (dataset fingerprint, foundation id, solver, lstsq rank) is recorded, and `ResolvedE0s.to_array(heads, z_table)` yields the `[n_heads, n_elements]` fp64 array legacy already uses.

4. **The loop** (`mace_torch/train/loop.py` + `ema.py`, `stage_two.py`, `schedulers.py`, `metrics.py`): explicit train/validate/checkpoint/EMA/stage-two, no Lightning, no callback framework. Terminology is **stage-two** (`start_stage_two`), the current name for what legacy spells SWA — and in develop the `--stage_two*` spellings are already option-string aliases on the `swa*` `dest`s (`mace/tools/arg_parser.py:801-822, 985-990`), so v1 keeps the new name and drops the old spelling. Multi-dataloader balancing goes to TRN-3 (#1577), composable losses/transforms to TRN-2 (#1576) and the LBFGS regime to TRN-5 (#1639) — TRN-1 (#1575) consumes a loss callable and dataloaders through the stage contract and ports only what the tiny task needs to train end-to-end with a mini-batch optimizer.

5. **Fine-tuning is not a special case.** The training section must not know whether it is fine-tuning: replay vs real head differ only in their data source (FT-1 (#1580) builds on this seam); minimal objects pass between sections.

6. **Layout and layering.** Code lives in `packages/mace-torch` under `mace_torch/train/` and the thin adapter `mace_torch/cli/run_train.py` (parse → `ResolvedConfig` → stages). Layer order `cli > train,data > models > nn > kernels > mace_core` holds: `nn`/`models` never import `train`, and no `__init__` eagerly imports `train` (that is what kills the god-module). `packages/` never imports the legacy `mace/`.

7. **Coexistence.** `mace_run_train` is owned by the launcher: `--engine legacy` (default) calls the frozen `mace.cli.run_train:run` **byte-identically**; `--engine v1` calls `mace_torch.cli.run_train`. Day-one behavior is unchanged; exposing `mace train` and the faithful legacy-flag port is CLI-1 (#1579).

8. **Open point this ticket settles (first PR):** the exact stage-boundary object schemas, and whether `DataStage` and `ModelStage` share a single `Statistics` object or recompute — decide and record it in the PR description.

**Task:**
1. Define the stage-boundary types in `mace_core` (`DataBundle`, `BuiltModel`, `TrainedModel`, statistics; Pydantic/dataclass, no torch), settling constraint 8.
2. Implement `DataStage` in `mace_torch`: dataset backends (DATA-1 (#1569)/2) + single graph build/collation (DATA-3 (#1571)), z-table, one `resolve_e0s` call with hard-error semantics, statistics (avg neighbors, energy/force mean+std) via the single `mace_core` implementation, dataloaders, and the callable pseudolabel seam.
3. Implement `ModelStage`: `ResolvedConfig` + statistics → `BaseMACE` + `MACEOutputs` through the ARCH-2 (#1561)/ARCH-3 (#1562) builders; declared observables create heads; backend resolved once at build time.
4. Implement `TrainStage`: the explicit loop (`train`/`train_one_epoch`/`evaluate` successors) with EMA, stage-two switch, checkpoint/resume hooks, and resolved-config-into-checkpoint-metadata — semantics ported from `mace/tools/train.py`, structure not copied.
5. Wire `mace_torch/cli/run_train.py` as the ~40-line orchestrator and register it under the launcher's `--engine v1` branch; `--engine legacy` stays byte-identical.
6. E2E CPU test: the tiny task trains under `--engine v1` — loss decreases, checkpoint + metadata written, resume from checkpoint works.

**Out of scope:** the config schema and `from_namespace` shim (CFG-1 (#1574)); composable losses, per-stage optimizer/LR/loss schedules, transform registry, model-transform hook (TRN-2 (#1576)); multi-dataloader balancing and metrics/logging/wandb (TRN-3 (#1577)); safetensors checkpoint format, DDP and SLURM ports (TRN-4 (#1578)); the L-BFGS full-batch regime (TRN-5 (#1639)); the `mace train/eval/...` CLI surface and legacy flag port (CLI-1 (#1579)); heads-as-modules, replay assembly, LoRA/freezing (FT-1 (#1580)); pseudolabel generation itself (FT-2 (#1581)); legacy-comparable final-error gating (GATE-3 (#1585)).

**Acceptance criteria:**
- [ ] `run()` successor is an orchestrator of ~40 lines wiring exactly three stages; no `argparse.Namespace` (and no `args.*` attribute) crosses a stage boundary — asserted by a test on the stage signatures.
- [ ] Stage boundary objects are typed and importable from `mace_core`; the shared-`Statistics`-vs-recompute decision is recorded.
- [ ] `resolve_e0s` is called exactly once, before model build; a z-table element with no resolvable E0 raises loudly (no `0.0` default, no all-zero lstsq fallback), and provenance is attached to the run.
- [ ] The loop is explicit — no Lightning/Accelerate, no callback/hook framework — and performs train/validate/checkpoint/EMA/stage-two; a resume from checkpoint continues training.
- [ ] The tiny task trains end-to-end on CPU under `--engine v1`: loss decreases and the written checkpoint carries the resolved config as metadata.
- [ ] `--engine legacy` output is byte-identical to before this PR (frozen path untouched).
- [ ] Layering holds: `nn`/`models` import no `train`; `packages/` imports no legacy `mace/` (import-linter green).

**Inventory gaps assigned here:**
- [ ] EMA affects evaluation — a smoke that would fail if `--ema` / `--ema_decay` silently did nothing. (`--ema_decay` is passed by many workflow tests via `tests/helpers.py`, but nothing asserts it changes a number.)
- [ ] `--dry_run` does what it claims: it is exercised (`tests/workflows/test_run_train_allkeys.py:430`, `test_run_train.py:1412,:1461`) but only as "does not crash"; assert that it returns at `mace/cli/run_train.py:1023-1025` before any epoch runs and writes no model.

**Verify:**
```bash
python -m pytest packages/mace-core/tests/test_stage_contracts.py -v        # typed boundaries, no-namespace assertion
python -m pytest packages/mace-torch/tests/train -v                          # loop: EMA, stage-two, checkpoint/resume, resolve_e0s-once
python -m pytest tests/workflows -k "v1 and tiny" -n 2                       # E2E: tiny task trains, loss decreases, resume works
```
**Review focus:** the stage-contract schemas — they freeze the boundary every follow-up (TRN-2 (#1576)/3/4/5, CLI-1 (#1579), FT-1 (#1580), ELEC-1 (#1591)) builds against — and the loop's semantic parity with legacy `train()` (EMA update ordering, the stage-two switch epoch, resume behavior), since GATE-3 (#1585) later gates on legacy-comparable losses/errors for this exact path.

---
